Add CouchDB support to the Notary

stefanberger opened the pull request:
This series of patches adds support for a CouchDB backend to the Notary server and signer. The code has been derived from the existing RethinkDB driver support and therefore resembles it in places.

The primary difference between the existing database backends and the CouchDB backend is that CouchDB does not support tables within databases. We work around this by encoding the name of tables into that of databases. A database `foo` and a table `bar` will then be used to form the database name `foo$bar`.

One other difference is that we want to be able to support the cloud version of a CouchDB, which is called 'Cloudant', as a backend. Here we may not have HTTPs client certificates for authentication but rely on username and password. Therefore, it is not required that a client certificate and key be provided to talk to the CouchDB backend. It is optional.

Otherwise the series of patches adds CouchDB related test cases to the server and signer and other parts of the notary code.

docker-jenkins commented:
Can one of the admins verify this patch?

stefanberger commented:
Can someone please kick the build. I failed due to other reasons...

stefanberger commented:
.. what I meant was, can someone restart the build since it failed due to reasons unrelated to the patches...

I know it's a lot of code, but can someone review this ?

docker-jenkins commented:
Can one of the admins verify this patch?

stefanberger commented:
I did not mean to close this pull request...

justincormack commented:
Needs a rebase.

stefanberger commented:
@justincormack I did that now along with some cleanups.

endophage commented:
Jenkins test this please.

ecordell commented:
@stefanberger Thanks for working on this, and I'm sorry it's been sitting here with no reviews.

I'm wary of merging support for another database type without a way to manage the complexity going forward. Currently, notary doesn't support all of the base spec of TUF let alone some of the features described in the TAPs. Adding support for the more TUF features will require model changes and migrations, and the support burden for those changes is multiplied by the number of database backends.

I think the ideal for this case would be to use go plugins to support third-party databases and move all non-sql databases outside of notary. That way the people with the domain expertise and requirements to run against a different database can support it separately from notary. 

@theupdateframework/notary-maintainers Any thoughts with this approach?

molepigeon commented:
I'm working in the team that is running Notary alongside IBM's container registry. We're using this driver already in production in order to run using Cloudant DBaaS as the backing database.

I appreciate your points about the support burden for migrations being increased. However, since my team and I will need to put in the work to accommodate database migrations in future anyway, we'd much rather do that work in the open with your blessing than have to do it internally.

Longer term, we'll be happy to talk through having more generic support for noSQL databases in particular, if that would help the project to support this and other noSQL drivers.

cyli commented:
@molepigeon @stefanberger  That'd be awesome if you wanted to maintain the nosql drivers :D   From a discussion with other maintainers, I think we'd still like to eventually move the code out of the repo over time, perhaps to a different repo under `theupdateframework` or basically to an "officially supported" repo of drivers, but out of the main notary repo, but since we have no immediate plans to implement a plugin system, that can be done at a later date and we can merge the CouchDB stuff in now after it's reviewed.

As for reviewing, apologies for the very long delay.  I think all the maintainers have been a bit slammed, and it's a largish change that we would like to do justice to.  I will try to have a look over the next week or so.

stefanberger commented:
@cyli Did you get a chance to review the code ?

cyli commented:
@stefanberger Apologies for the delay, I'm a little more than halfway through - mostly just some tests left to look at.  I'll try to get it done soon, although I'll be out in the next week for a conference.

cyli commented on server/storage/couchdb.go:
```diff
@@ -0,0 +1,591 @@
+
+	_, err = db.Put(context.TODO(), id, file)
+	if err != nil {
+		return ErrOldVersion{}
```
Hmm... more error parsing may be needed.  It seems like there may be other reasons a `db.Put` would error other than a conflict, which is what would signify an old version.

stefanberger replied:
I am parsing the returned error string now and return err if it doesn't match, ErrOldVersion() otherwise.

cyli commented on server/storage/couchdb.go:
```diff
@@ -0,0 +1,591 @@
+
+	id := genID(gun.String(), update.Role.String(), update.Version)
+
+	row, _ := db.Get(context.TODO(), id)
```
I'm not familiar with couch, but rethink basically has an insert but error on conflict, which would mean that we can't get a race where we'd get first, see no conflict, and then insert.  Is there a similar mechanism for CouchDB?  Is that what the rev # does?  (e.g. does the update fail if the latest rev is higher than the rev we attempt to update with?)  Otherwise it seems like 2 updates could race here.

stefanberger replied:
The _rev field is the revision of a document, which is some sort of hash over its contents. To updated it, one needs to show the current version, which then automatically changes to another value once the update is done. This excludes concurrent updates and races. 

https://writing.jan.io/2013/12/19/understanding-couchdb-conflicts.html

cyli replied:
Thank you!  I notice a statement at the bottom stating that if writes happen on two separate CouchDB instances in a cluster, there could be a conflict as well - this blog post seems to suggest only writing to one node by using a load balancer.

However, in a mult-notary server scenario, where each node is speaking to a cluster of CouchDB nodes, this may not make sense.

Does CouchDB have read/write quorums?  I am not entirely sure from reading.  It seems like maybe Cloudant does?  (https://danielwertheim.se/how-to-customize-quorum-with-cloudant-using-mycouch/) But it seems like by default CouchDB may not?  (http://grokbase.com/t/couchdb/dev/153rwp8j3x/could-couchdb-2-0-fix-actual-read-quorum).  If it does, I think we should specify read/write quorums.

From http://guide.couchdb.org/draft/conflicts.html it seems like if two incoming writes happen, both documents will be saved on the server but they will be in conflict.  When returning the current version of the document, one of those conflicting documents will be chosen, and the choice will be the same across all the nodes.

My concern is that if two concurrent writes happen:

1. There will be no error, so both writes will seem to succeed, although in practice only one of those writes will be surfaced since there does not seem to be code in this PR to handle conflicting documents.  I'm not sure what that should look like, even.

2. An update consists of multiple metadata files - this means that the current root may not match the current timestamp if there were a conflict writing either the timestamp or the root metadata, or both.  The timestamp from update A could be chosen as the current document by CouchDB, and the root from update B be chosen as the current document by CouchDB.  This may cause issues when pulling if there is a conflict that results in the current root metadata not matching the current timestamp metadata.  The update logic (let's skip the case where we have any cached metadata for now) is as follows:

    1. Download the current root metadata
    1. Download the current timestamp metadata and verify that it was signed by the right key.
    1. Download the rest of the metadata, do validations, etc...

    If the conflict I mentioned above happened (timestamp from A, root from B), then the client would not be able to update any metadata until the conflict(s) were resolved on the server.

In RethinkDB, this is resolved by using read and write quorums.  2 simultaneous writes can't happen - whomever writes the timestamp first wins the race (because if there is a conflict on timestamp, everything else is rolled back).  And getting the current version of anything or getting a particular version number requires a read quorum.

However, without write quorum, a "transaction" of metadata won't be rolled back if there's a timestamp conflict, and that could mess up `GetCurrent` and `GetVersion` on the server.

stefanberger replied:
Here's the configuration of a test database on cloudant.com:

curl https://64b40286-4fad-4454-b691-b0f1b0907376-bluemix:${password}@64b40286-4fad-4454-b691-b0f1b0907376-bluemix.cloudant.com/test

```
{"update_seq":"0-g1AAAAHveJzLYWBg4MhgTmFQTUlKzi9KdUhJMjTSS8rVTU7WTStK1DUw1EvOyS9NScwr0ctLLckBKmdKZEiS____f1YiA0ijClyjBQF9SQpAMskeoZUEO5McQHrjybI2AaS1Hqu1hgT05rEASYYGIAXUPp9kqyHaF0C078e03oAo_Qcg-u-Taf0DiHaQ77MAWOqd2w","db_name":"test","sizes":{"file":1468,"external":0,"active":0},"purge_seq":0,"other":{"data_size":0},"doc_del_count":0,"doc_count":0,"disk_size":1468,"disk_format_version":6,"data_size":0,"compact_running":false,"cluster":{"q":8,"n":3,"w":2,"r":2},"instance_start_time":"0"}
```

There are 3 nodes (n=3) and read quorum is '2' ("r":2) and the write quorum is also '2' ("w":2). I didn't have to set this up as part of the database creation but it was configured like this automatically. So presumably they do w = r = round_down(n / 2) + 1, which results in w + r > n for n odd. So consistency is guaranteed for Cloudant.

I have had my challenges trying to set up a CouchDB cluster using containers on different machines...

stefanberger replied:
From some experiments I have done now with CouchDB clustering setup it looks like `r` and `w` are automatically adjusted to `round_down(n/2) + 1` when new nodes are added. We should be fine if now every read and write is happening with these adjusted quorum values.

cyli replied:
Awesome, thank you for checking.

stefanberger replied:
The way I understand it is that CouchDB's and Cloudant's internal configuration enforces read and write consistency. The documentation is here: http://docs.couchdb.org/en/2.1.1/api/database/common.html#get--db

However, the frequently used 'find' operation doesn't enforce read consistency but uses a quorum of '1' instead: http://docs.couchdb.org/en/2.1.1/api/database/find.html#post--db-_find

stefanberger replied:
@cyli Following this, should we adjust the read quorum value for the 'find' operation to the proper value to ensure consistency? My feeling is if values returned from the 'find' operation are returned to the user while the data is changing, the consistency wouldn't matter as much than in cases where those data are used internally for further updates, deletes, etc.

cyli commented on server/storage/storage_test.go:
```diff
@@ -434,6 +434,9 @@ func testGetChanges(t *testing.T, s MetaStore) {
 	if _, ok := s.(RethinkDB); ok {
 		require.Error(t, err1)
 		require.Error(t, err2)
+	} else if _, ok := s.(CouchDB); ok {
```
It might be clearer to convert this `if/else if/else` block at this point to a `switch` statement based on the type of `s`, at this point.  E.g.:

```golang
switch s.(type) {
case RethinkDB: ...
case CouchDB: ...
default: ...
}
```

stefanberger replied:
Fixed.

cyli commented on signer/keydbstore/couch_keydbstore.go:
```diff
@@ -0,0 +1,434 @@
+		doc.ID.ID = keyID
+		if _, err := db.Put(context.TODO(), keyID, doc); err != nil {
+			return err
+		}
```
Non-blocking:  there should only be 1 key with this key ID I think - should this return after?

justincormack commented:
The model that eg `containerd` uses for optional components is that they live in their own repo in the org and there is a build tag or similar that will build them, while the repo can have its own CI that tests the specifics. eg https://github.com/containerd/zfs which is not built or used by default. This keeps the core focused but allows a combined build relatively easily.

stefanberger commented on utils/configuration.go:
```diff
@@ -178,6 +190,62 @@ func ParseRethinkDBStorage(configuration *viper.Viper) (*RethinkDBStorage, error
+	if dburl.User != nil {
+		store.Username = dburl.User.Username()
+		store.Password, _ = dburl.User.Password()
+		// remove username and password from URL so it does not appear in logs
```
I removed the line and comment since dburl will not be used any further in this function:
```
dburl.User = nil
```

stefanberger commented on storage/couchdb/bootstrap.go:
```diff
@@ -0,0 +1,106 @@
+		}
+	}
+
+	// If the password is empty, pass false to the password parameter
```
I think it's not relevant. Removed it.

stefanberger commented on storage/couchdb/couchdb.go:
```diff
@@ -0,0 +1,157 @@
+	if err != nil {
+		return nil, err
+	}
+	// We have to use the original CipherSuites to be able to create a
```
It's been a while... I remember having connectivity problems to either CouchDB or Cloudant, not sure. But somehow the CipherSuites were not correct or missing entries that the other side could accept.

stefanberger replied:
I added documentation to the top of this script that talks about this.

stefanberger replied:
When I disable the line

'''
       // We have to use the original CipherSuites to be able to create a
        // connection with CouchDB
        // t.CipherSuites = tls.Config{}.CipherSuites
'''

I get the following error in the logs:
```
server_1  | {"level":"fatal","msg":"Error starting couchdb driver: Get https://cdb-01.cdb:6984/_session: remote error: tls: insufficient security level","time":"2018-05-21T16:31:44Z"}
```

stefanberger replied:
Server side reports this here:

```
SSL: hello: tls_handshake.erl:116:Fatal error: insufficient security
```

On the golang side:

tls.Config{}.CipherSuites seems to set the following:

			TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305,
			TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305,
			TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,
			TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,
			TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,
			TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,

(non-hardware case from https://golang.org/src/crypto/tls/common.go?s=12277:20617#L912)

whereas we would otherwise only get:

	tls.TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,
	tls.TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,
	tls.TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,
	tls.TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,

(form https://github.com/docker/go-connections/blob/master/tlsconfig/config_client_ciphers.go)

so the difference seems to be:

			TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305,
			TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305,

The failure seems to be caused by the server, which is written in erlang. It may be related to some erlang regression:

https://stackoverflow.com/questions/36784391/rabbitmq-3-6-1-erlang-18-3-tls-insufficient-security-failures

cyli replied:
Thanks for clarifying.  We should probably add the CHACHA20 ciphers to the list of supported ciphers in go-connections since they are supported for TLS 1.3 - cc @endophage.

stefanberger commented on storage/couchdb/couchdb.go:
```diff
@@ -0,0 +1,157 @@
+
+	rows, err := db.AllDocs(context.TODO())
+	if err != nil {
+		return nil, nil, fmt.Errorf("GetAllDocs: Could not get AllDocs: %s", err)
```
Fixed. Let me push now with some small patches on top of the existing PR. I want to see how this all reacts here.

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/moby/moby/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "couchdb" git@github.com:stefanberger/notary.git somewhere
$ cd somewhere
$ git rebase -i HEAD~842353914768
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

stefanberger commented:
I get some odd errors for ci/circleci. Will keep it entertained for a while...

stefanberger commented on storage/couchdb/couchdb.go:
```diff
@@ -0,0 +1,157 @@
+
+	err = rows.Err()
+	if err != nil {
+		return nil, nil, fmt.Errorf("GetAllDocs: Error occurred: %s", err)
```
I modified it to that now.

stefanberger commented on storage/couchdb/bootstrap.go:
```diff
@@ -0,0 +1,106 @@
+func makeDB(client *kivik.Client, name string) error {
+	err := client.CreateDB(context.TODO(), name)
+	if err != nil {
+		exists, nerr := client.DBExists(context.TODO(), name)
```
I added a test case now that tests the err != nil and returns nil as an error for a database that's being created again.

cyli replied:
Thank you!  Sorry for being so obtuse - I'm sure it's here, but I can't seem to spot this case.  Would you mind pointing it out to me, because I think I'm blind.  :)

stefanberger replied:
The test is here: https://github.com/theupdateframework/notary/blob/61c31cafe4211b8c67cc78fae3895e4538495470/server/storage/couch_realdb_test.go#L87

stefanberger commented on couchdb/couchdb_startup.sh:
```diff
@@ -0,0 +1,87 @@
+#!/usr/bin/env bash
```
I used this script because it allows us to run the server and signer with less privileged users. The less privileged users cannot create a database themselves but can use a database setup by an admin. I get the following type of an error otherwise. So, yeah, its scripted bootstrapping of the database tables for the purpose of privilege separation.
```
{"level":"fatal","msg":"unable to create DB notarysigner$private_keys: Unauthorized: You are not a server admin.","time":"2018-05-21T15:02:13Z"}
```

The privilege separation was mainly for CouchDB. For Cloudant, iirc, we cannot setup other users with their own credentials but have to use the account given to use by the cloud.

stefanberger replied:
I added the reason for this script to the top of the file including documentation regarding the environment variables that are needed.

cyli replied:
In https://github.com/theupdateframework/notary/pull/1279#discussion_r195611624 I mentioned that maybe the admin connection and the user connection should be the same.  I'm going to amend that by saying that maybe when establishing the admin connection for bootstrapping (which can call the same function), maybe we can check for the admin user/passphrase environment variable?

If I am understanding everything correctly, currently the URL that is passed means that all connections are made with the server/signer user and passphrase, which means that you can't establish an admin connection.  But if you got the username and passphrase differently if `bootstrap` is set to true, then when doing the bootstrap you would be able to connect as an administrator and generate the tables, etc. via the code that has already been written.

stefanberger replied:
The environment variable to get the admin user/passphrase from is DBURL ?

If we go through the bootstrap we would need a way to pass two usernames (signer and server) for the two databases along with two passwords. How do we do that?

stefanberger replied:
MySQL/PostreSQL are initialized using SQL scripts in `notarysql/mysql-initdb.d/initial-notaryserver.sql` for example. There the notaryserver database is created as well as the server user. The server user is granted access to the notaryserver database. 
```
CREATE DATABASE IF NOT EXISTS `notaryserver`;

CREATE USER "server"@"%" IDENTIFIED BY "";

GRANT
        ALL PRIVILEGES ON `notaryserver`.* 
        TO "server"@"%";
```

The SQL files are run by the MySQL database container where these files are mapped into the container's `docker-entrypoint-initdb.d` directory. 

So I would justify this script here to be an equivalent of the MySQL bootstrapping with a couple of differences:
- there's no built-in bootstrapping mechanism in the CouchDB container that would run some setup scripts like those SQL scripts mapped into a certain directory; so we manipulate the `entrypoint` and `command` of the CouchDB container
- we cannot create databases in CouchDB but have to create tables instead and grant access to these tables. 
- we set the passwords for the signer and server users here so we don't have to provide the admin password to the server and signer containers for the bootstrapping.
- CouchDB needs a couple of non-Notary databases setup during bootstrapping: `_users`, `_replicator`, `_global_changes`. These are described here: http://docs.couchdb.org/en/latest/install/setup.html

stefanberger commented on storage/couchdb/couchdb.go:
```diff
@@ -0,0 +1,157 @@
+	return client, nil
+}
+
+// AdminConnection sets up an admin CouchDB connection to the host (`host:port` format)
```
I shamelessly recycled that from RethinkDB's usage in cmd/notary-server/config.go:

```
                if doBootstrap {
                        sess, err = rethinkdb.AdminConnection(tlsOpts, storeConfig.Source)
                } else {
                        sess, err = rethinkdb.UserConnection(tlsOpts, storeConfig.Source, storeConfig.Username, storeConfig.Password)
                }
```

cyli replied:
Ah ok, for RethinkDB that is because if no username is supplied the admin user is used, and by default the admin user has no password.  This seems problematic.  I've filed https://github.com/theupdateframework/notary/issues/1366 to fix this.  If both functions need to take a username and password, it'd probably be better just to have a `Connection` function in the `couchdb` package.

stefanberger commented on server/storage/couchdb.go:
```diff
@@ -0,0 +1,591 @@
+	return ChangefeedTableName
+}
+
+// unmarshal in an anonymous struct
```
Correct. It looks like it can be removed.

```
diff --git a/server/storage/couchdb.go b/server/storage/couchdb.go
index bc86a608..da141ff9 100644
--- a/server/storage/couchdb.go
+++ b/server/storage/couchdb.go
@@ -49,38 +49,6 @@ func (cdb CChange) TableName() string {
 	return ChangefeedTableName
 }
 
-// unmarshal in an anonymous struct
-func cdbTUFFileFromJSON(data []byte) (interface{}, error) {
-	a := struct {
-		CreatedAt  time.Time `json:"created_at"`
-		UpdatedAt  time.Time `json:"updated_at"`
-		DeletedAt  time.Time `json:"deleted_at"`
-		Gun        string    `json:"gun"`
-		Role       string    `json:"role"`
-		Version    int       `json:"version"`
-		SHA256     string    `json:"sha256"`
-		Data       []byte    `json:"data"`
-		TSchecksum string    `json:"timestamp_checksum"`
-	}{}
-	if err := json.Unmarshal(data, &a); err != nil {
-		return CDBTUFFile{}, err
-	}
-	return CDBTUFFile{
-		Timing: couchdb.Timing{
-			CreatedAt: a.CreatedAt,
-			UpdatedAt: a.UpdatedAt,
-			DeletedAt: a.DeletedAt,
-		},
-		GunRoleVersion: []interface{}{a.Gun, a.Role, a.Version},
-		Gun:            a.Gun,
-		Role:           a.Role,
-		Version:        a.Version,
-		SHA256:         a.SHA256,
-		Data:           a.Data,
-		TSchecksum:     a.TSchecksum,
-	}, nil
-}
-
 func cdbChangeFromJSON(data []byte) (interface{}, error) {
 	res := CChange{}
 	if err := json.Unmarshal(data, &res); err != nil {
diff --git a/server/storage/couchdb_models.go b/server/storage/couchdb_models.go
index fd2a2b7e..5d98fc74 100644
--- a/server/storage/couchdb_models.go
+++ b/server/storage/couchdb_models.go
@@ -17,7 +17,6 @@ var (
 			"6": []string{"gun", "created_at"},
 			"7": []string{"_id", "created_at"},
 		},
-		JSONUnmarshaller: cdbTUFFileFromJSON,
 	}
 
 	// ChangeCouchTable is the table definition for changefeed objects

```

eiais commented:
It looks like you added a commit from #1353 in your most recent commit.

stefanberger commented:
@eiais Rebased on master again and it's gone. Not sure how it got into the middle of this.

stefanberger commented on storage/couchdb/bootstrap.go:
```diff
@@ -0,0 +1,106 @@
+func SetupDB(client *kivik.Client, dbName string, tables []Table) error {
+	var err error
+
+	if tables == nil {
```
I removed that now. Only a test case was calling it like that.

stefanberger commented on storage/couchdb/couchdb.go:
```diff
@@ -0,0 +1,157 @@
+	}
+
+	if username != "" {
+		var basicAuth chttp.Authenticator = &chttp.BasicAuth{
```
For the setting of the transport I followed an example shown in the Kivik library:

https://github.com/go-kivik/couchdb/blob/master/auth.go#L41

```
// Example:
//
//     setXport := couchdb.SetTransport(&http.Transport{
//         // .. custom config
//     })
//     client, _ := kivik.New( ... )
//     client.Authenticate(setXport)
```

The client.Authenticate(setXport) doesn't seem to do much except for setting the transport:

First: https://github.com/go-kivik/couchdb/blob/master/auth.go#L17
Then: https://github.com/go-kivik/couchdb/blob/master/auth.go#L33

The 2nd authenticate in my code is this one:

```               if err = client.Authenticate(context.TODO(), basicAuth); err != nil {
                        return nil, err
                }
```

This is the implementation of the Authenticate() function:

https://github.com/go-kivik/couchdb/blob/master/chttp/auth.go#L45

It sets the transport at the end for future use: https://github.com/go-kivik/couchdb/blob/master/chttp/auth.go#L71

I will leave some comments in the code regarding that.
The function then returns a client object that is set up to use for accessing the DB.

I can remove the check for username to require basic authentication all the time, not just client cert auth.

cyli replied:
Huh, this seems like a unintuitive interface decision - thank you very much for documenting it.  I don't feel particularly strongly about the check for the username - is it invalid to connect to a CouchDB without basic auth?  If so, then I think it makes sense to fail if there is no username.  Otherwise I am ok with this as is.

stefanberger replied:
The CouchDB container accepts the `COUCHDB_USER` and `COUCHDB_PASSWORD` parameters but they are not mandatory. In case they are omitted, the CouchDB can be accessed without any authentication.

https://hub.docker.com/_/couchdb/

cyli replied:
Ah ok thanks for explaining.  I'm ok with this snippet as is, thank you!

stefanberger commented:
@cyli Did I address the comments blocking a merge ?

stefanberger commented on fixtures/couchdb/ca.pem:
```diff
@@ -0,0 +1,18 @@
+-----BEGIN CERTIFICATE-----
```
I extended `regenerateTestingCerts.sh` now to do that for the CouchDB test certs.

cyli replied:
Thanks!

cyli commented:
@stefanberger Apologies again, I will review the changes as soon as I get a chance!  Thanks for being so responsive!

stefanberger commented:
@cyli That would be great. I should probably squash some of the patches that remove the anonymous structure usage into the patches where this was introduced... Let me know.

cyli commented:
(Apologies still looking through the changes - more comments soon)

cyli commented:
Sorry this took so long @stefanberger.  I have a couple of nitpicks about the CouchDB bootstrapping, but mainly I still have some concerns about consistent reads and writes. Again, I am super unfamiliar with CouchDB, so maybe it is not actually a problem, but that is the main blocker, if that could be addressed.  :). 

Thanks again for your patience and perseverance!

stefanberger commented:
@cyli I added some more patches on top of this now. One improves query performance on a production system by changing the selector of the query and avoiding sorting the query by the database server. 

Another request from the production team was to have the deletedAt field omitted from documents when it was showing dates like `0001-01-01T00:00:00Z`, which indicated that this field wasn't set. So I converted time.Time to *time.Time along with the necessary changes in a few places in the code and test case to accommodate this. I suppose this is ok since it is database internal.

Further, I fixed a test case when used with a Cloudant. The test case tries to do access a database with a user account but a Cloudant account not allow additional user accounts to be created. So I added a patch to the test case that handles a lower level error chttp.HTTPError{} from the CouchDB driver, which is acceptable in this case., but trickles up into the test code. All other tests are passing fine and are worth being able to run with a Cloudant as well.

stefanberger commented:
@cyli In the RethinkDB driver you seem to be writing with write_acks set to majority. Reading seems to also use the default of 'single', so no consistency is guaranteed. The only time consistency is a concern is in GetChanges(). Is this correct?

stefanberger commented:
Can merge?
